pkg/pillar/types/errortime.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2020 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package types
+
+import (
+	"reflect"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// Common error with timestamp
+
+// ErrorAndTime is used by many EVE agents
+type ErrorAndTime struct {
+	Error     string
+	ErrorTime time.Time
+}
+
+// SetErrorNow uses the current time
+func (etPtr *ErrorAndTime) SetErrorNow(errStr string) {
+	etPtr.Error = errStr
+	etPtr.ErrorTime = time.Now()
+}
+
+// SetError is when time is specified
+func (etPtr *ErrorAndTime) SetError(errStr string, errorTime time.Time) {
+	etPtr.Error = errStr
+	etPtr.ErrorTime = errorTime
+}
+
+// ClearError removes it
+func (etPtr *ErrorAndTime) ClearError() {
+	etPtr.Error = ""
+	etPtr.ErrorTime = time.Time{}
+}
+
+// ErrorAndTimeWithSource has an additional field "ErrorSourceType"
+// which is used to selectively clear errors by calling IsErrorSource before
+// calling ClearErrorWithSource. See zedmanager and volumemgr for example use.
+type ErrorAndTimeWithSource struct {
+	ErrorSourceType interface{}
+	Error           string
+	ErrorTime       time.Time
+}
+
+// SetError - Sets error state with no source type
+func (etsPtr *ErrorAndTimeWithSource) SetError(errStr string, errTime time.Time) {
+	etsPtr.Error = errStr
+	etsPtr.ErrorSourceType = nil
+	etsPtr.ErrorTime = errTime
+}
+
+// SetErrorWithSource - Sets error state. Source needs to be a type
+func (etsPtr *ErrorAndTimeWithSource) SetErrorWithSource(errStr string,
+	source interface{}, errTime time.Time) {
+
+	if !allowedSourceType(source) {
+		log.Fatalf("Bad ErrorSourceType %T", source)
+	}
+	etsPtr.Error = errStr
+	etsPtr.ErrorSourceType = source
+	etsPtr.ErrorTime = errTime
+}
+
+// IsErrorSource returns true if the source type matches
+func (etsPtr *ErrorAndTimeWithSource) IsErrorSource(source interface{}) bool {
+	if !allowedSourceType(source) {
+		log.Fatalf("Bad ErrorSourceType %T", source)
+	}
+	if etsPtr.Error == "" {
+		return false
+	}
+	return reflect.TypeOf(source) == reflect.TypeOf(etsPtr.ErrorSourceType)
+}
+
+// ClearErrorWithSource - Clears error state
+func (etsPtr *ErrorAndTimeWithSource) ClearErrorWithSource() {
```

Comment:
If you make ErrorAndTime and embedded field, you can just rename this function as "ClearError"
The prefix "WithSource" doesn;t add much value..

type ErrorAndTime struct {
<snip>
}

func (etPtr *ErrorAndTime) ClearError() {
 <snip>
}

type ErrorAndTimeWithSource struct {
ErrorAndTime
Source interface{}
}

func (etsPtr *ErrorAndTimeWithSource) ClearError() {. <--- Note --Same name as the fn of embedded field
   etsPtr.ErrorAndTime.ClearError()
   etsPtr.Source = nil
}

func (etPtr *ErrorAndTime) ClearError() {
 <snip>
}

Reply:
The issue I see is that the embeeded object (ErrorAndTIme vs. ErrorAndTimeWIthSource) are completely hidden from the code which calls, which is why it made sense to me to use different function names for the two objects. Without that it would be easier to confuse. This means not embedding ErrorAndTime inside ErrorAndTimeWithSource.